data-plane/dispatcher/src/main/java/dev/knative/eventing/kafka/broker/dispatcher/impl/consumer/CloudEventOverridesMutator.java

```diff
@@ -28,21 +29,42 @@ public class CloudEventOverridesMutator implements CloudEventMutator {
 
     private final DataPlaneContract.CloudEventOverrides cloudEventOverrides;
 
+    private static final CloudEventDeserializer cloudEventDeserializer = new CloudEventDeserializer();
+
     public CloudEventOverridesMutator(final DataPlaneContract.CloudEventOverrides cloudEventOverrides) {
         this.cloudEventOverrides = cloudEventOverrides;
     }
 
     @Override
-    public CloudEvent apply(CloudEvent cloudEvent) {
-        if (cloudEventOverrides.getExtensionsMap().isEmpty()) {
-            return cloudEvent;
-        }
+    public CloudEvent apply(ConsumerRecord<Object, CloudEvent> record) {
+        final var cloudEvent = this.maybeDeserializeFromHeaders(record);
         final var builder = CloudEventBuilder.from(cloudEvent);
+        applyKafkaMetadata(builder, record.partition(), record.offset());
         applyCloudEventOverrides(builder);
         return builder.build();
     }
 
+    private CloudEvent maybeDeserializeFromHeaders(ConsumerRecord<Object, CloudEvent> record) {
+        if (record.value() != null) {
+            return record.value();
+        }
+        // A valid CloudEvent in the CE binary protocol binding of Kafka
+        // might be composed by only Headers.
+        //
+        // KafkaConsumer doesn't call the deserializer if the value
+        // is null.
+        //
+        // That means that we get a record with a null value and some CE
+        // headers even though the record is a valid CloudEvent.
+        return cloudEventDeserializer.deserialize(record.topic(), record.headers(), null);
+    }
+
     private void applyCloudEventOverrides(CloudEventBuilder builder) {
         cloudEventOverrides.getExtensionsMap().forEach(builder::withExtension);
     }
+
+    private void applyKafkaMetadata(CloudEventBuilder builder, Number partition, Number offset) {
+        builder.withExtension("knativekafkapartition", partition);
+        builder.withExtension("knativekafkaoffset", offset);
```

Comment:
can we make those `private static final String` ?

Reply:
I understand the prefix of knative...  but wondering if the actual upstream/spec, would want to say something on the metadata in https://github.com/cloudevents/spec/blob/main/cloudevents/bindings/kafka-protocol-binding.md  ?

Reply:
I didn't see anything in particular while reading the docs there, I'll ask at the serverless WG meeting tomorrow